py-client/client_api.py

```diff
@@ -88,30 +90,48 @@ def model_list(self, label=None):
         response = response.decode('utf-8') if isinstance(response, bytes) else response
         return json.loads(response)
 
-    def segmentation(self, model, image_in, image_out):
+    def segmentation(self, model, image_in, image_out, save_doc=False):
         """
         3D image segmentation using segmentation method
         :param model: model name according to the output of model_list()
         :param image_in: input 3D image file name
         :param image_out: output files will be stored
-        :return: returns json containing extreme points for the segmentation mask
+        :param save_doc: save input image in server for future reference; server will return doc id in params
+        :return: returns params containing extreme points for the segmentation mask and other info
 
         Output 3D binary mask will be saved to the specified file
         """
         logger = logging.getLogger(__name__)
         logger.debug('Preparing for Segmentation Action')
 
         selector = '/' + self.api_version + '/segmentation?model=' + AIAAUtils.urllib_quote_plus(model)
+        selector += '&save_doc=' + ('true' if save_doc else 'false')
+        if self.doc_id:
+            selector += '&doc=' + AIAAUtils.urllib_quote_plus(self.doc_id)
+
         fields = {'params': '{}'}
-        files = {'datapoint': image_in}
+        files = {'datapoint': image_in} if self.doc_id is None else {}
 
         logger.debug('Using Selector: {}'.format(selector))
         logger.debug('Using Fields: {}'.format(fields))
         logger.debug('Using Files: {}'.format(files))
 
         form, files = AIAAUtils.http_post_multipart(self.server_url, selector, fields, files)
+        form = json.loads(form) if isinstance(form, str) else form
+
+        params = form.get('params')
+        if params is None:
+            points = json.loads(form.get('points'))
+            params = {'points': (json.loads(points) if isinstance(points, str) else points)}
+        else:
+            params = json.loads(params) if isinstance(params, str) else params
+
         AIAAUtils.save_result(files, image_out)
-        return form
+
+        if self.doc_id is None:
+            self.doc_id = form.get('doc') if isinstance(form, dict) else None
```

Comment:
is this the check we want?
I thought after Line 120, form should be dict?
and is form guaranteed to have 'doc'?

Reply:
Fixed